src/devices/Mpu9250/Mpu6500.cs

```diff
@@ -0,0 +1,899 @@
+        /// </summary>
+        public uint FifoCount
+        {
+            get
+            {
+                Span<byte> rawData = stackalloc byte[2] { 0, 0 };
+                ReadBytes(Register.FIFO_COUNTH, rawData);
+                return BinaryPrimitives.ReadUInt16BigEndian(rawData);
+            }
+        }
+
+        /// <summary>
+        /// Get or set the FIFO (First In First Out) modes
+        /// </summary>
+        public FifoModes FifoModes
+        {
+            get { return (FifoModes)(ReadByte(Register.FIFO_EN)); }
+            set
+            {
+                if (value != FifoModes.None)
+                {
+                    // Make sure the FIFO is enabled
+                    var usrCtl = (UserControls)ReadByte(Register.USER_CTRL);
+                    usrCtl = usrCtl | UserControls.FIFO_RST;
+                    WriteRegister(Register.USER_CTRL, (byte)usrCtl);
+                }
+                else
+                {
+                    // Deactivate FIFO
+                    var usrCtl = (UserControls)ReadByte(Register.USER_CTRL);
+                    usrCtl = usrCtl & ~UserControls.FIFO_RST;
+                    WriteRegister(Register.USER_CTRL, (byte)usrCtl);
+                }
+                WriteRegister(Register.FIFO_EN, (byte)value);
+            }
+        }
+
+        /// <summary>
+        /// Read data in the FIFO (First In First Out) buffer, read as many data as the size of readData byte span
+        /// You should read the number of data available in the FifoCount property then
+        /// read them here.
+        /// You will read only data you have selected in FifoModes.
+        /// Data are in the order of the Register from 0x3B to 0x60.
+        /// ACCEL_XOUT_H and ACCEL_XOUT_L
+        /// ACCEL_YOUT_H and ACCEL_YOUT_L 
+        /// ACCEL_ZOUT_H and ACCEL_ZOUT_L
+        /// TEMP_OUT_H and TEMP_OUT_L
+        /// GYRO_XOUT_H and GYRO_XOUT_L
+        /// GYRO_YOUT_H and GYRO_YOUT_L
+        /// GYRO_ZOUT_H and GYRO_ZOUT_L
+        /// EXT_SENS_DATA_00 to EXT_SENS_DATA_24
+        /// </summary>
+        /// <param name="readData">Data which will be read</param>
+        public void ReadFifo(Span<byte> readData)
+        {
+            ReadBytes(Register.FIFO_R_W, readData);
+        }
+
+        #endregion
+
+        #region Calibration and tests
+
+        /// <summary>
+        /// Perform full calibration the gyroscope and the accelerometer
+        /// It will automatically adjust as well the offset stored in the device
+        /// The result bias will be stored in the AcceloremeterBias and GyroscopeBias
+        /// </summary>
+        public (Vector3 gyroscopeBias, Vector3 AccelerometerBias) CalibrateGyroscopeAccelerometer()
+        {
+            // = 131 LSB/degrees/sec
+            const int GyroSensitivity = 131;
+            // = 16384 LSB/g
+            const int AccSensitivity = 16384;
+            byte i2cMaster;
```

Comment:
this method should be split into calibration and setting calibration params

Reply:
What do you mean?  The function does the calibration for both the gyroscope and the accelerometer. There are no real settings needed for the calibration. It's done by filling the FIFO with readings.

Reply:
I meant to make the function shorter by splitting it into two parts and make it slightly easier to read